File: ios/LoadcastCore/Sources/LoadcastCore/BodyFatigue.swift
/// 종목 → 부위 그룹 매핑과 부위별 피로 추정. docs/SPEC.md §7.
public enum BodyFatigue {

    public static let typeToGroups: [WorkoutType: Set<BodyGroup>] = [
        .crossfit: [.lower, .upper, .cardio],
        .running: [.lower, .cardio],
        .weight: [.lower, .upper, .core],
        .yoga: [.core],
        .swimming: [.upper, .cardio],
        .cycling: [.lower, .cardio],
        .hiking: [.lower, .cardio],
        .other: [.cardio],
    ]

    /// 최근 3일 [today-2, today] 그룹별 부하 합을 3구간으로 버킷링.
    public static func estimate(records: [WorkoutRecord], today: DayDate) -> [BodyGroup: FatigueLevel] {
        let from = today.adding(days: -2)
        var sums: [BodyGroup: Int] = Dictionary(uniqueKeysWithValues: BodyGroup.allCases.map { ($0, 0) })
        for record in records where record.date >= from && record.date <= today {
            for group in typeToGroups[record.type] ?? [] {
                sums[group, default: 0] += record.load
            }
        }
        return sums.mapValues(bucket)
    }

    private static func bucket(_ load: Int) -> FatigueLevel {
        switch load {
        case ..<100: .low
        case ..<250: .moderate
        default: .high
        }
    }
}

File: android/core/src/main/kotlin/com/loadcast/core/BodyFatigue.kt
package com.loadcast.core

import java.time.LocalDate

/** 종목 → 부위 그룹 매핑과 부위별 피로 추정. docs/SPEC.md §7. */
object BodyFatigue {

    val typeToGroups: Map<WorkoutType, Set<BodyGroup>> = mapOf(
        WorkoutType.CROSSFIT to setOf(BodyGroup.LOWER, BodyGroup.UPPER, BodyGroup.CARDIO),
        WorkoutType.RUNNING to setOf(BodyGroup.LOWER, BodyGroup.CARDIO),
        WorkoutType.WEIGHT to setOf(BodyGroup.LOWER, BodyGroup.UPPER, BodyGroup.CORE),
        WorkoutType.YOGA to setOf(BodyGroup.CORE),
        WorkoutType.SWIMMING to setOf(BodyGroup.UPPER, BodyGroup.CARDIO),
        WorkoutType.CYCLING to setOf(BodyGroup.LOWER, BodyGroup.CARDIO),
        WorkoutType.HIKING to setOf(BodyGroup.LOWER, BodyGroup.CARDIO),
        WorkoutType.OTHER to setOf(BodyGroup.CARDIO),
    )

    /** 최근 3일 [today-2, today] 그룹별 부하 합을 3구간으로 버킷링. */
    fun estimate(records: List<WorkoutRecord>, today: LocalDate): Map<BodyGroup, FatigueLevel> {
        val from = today.minusDays(2)
        val sums = BodyGroup.entries.associateWithTo(LinkedHashMap()) { 0 }
        records.filter { it.date in from..today }.forEach { record ->
            typeToGroups.getValue(record.type).forEach { group ->
                sums[group] = sums.getValue(group) + record.load
            }
        }
        return sums.mapValues { (_, load) -> bucket(load) }
    }

    private fun bucket(load: Int): FatigueLevel = when {
        load < 100 -> FatigueLevel.LOW
        load < 250 -> FatigueLevel.MODERATE
        else -> FatigueLevel.HIGH
    }
}
